packages/contracts/contracts/L1/ZKTrieHasher.sol

```diff
@@ -75,31 +75,65 @@ contract ZKTrieHasher {
     }
 
     /**
-     * @notice Computes a hash of elements.
+     * @notice Computes a root hash of elements tree.
      *
      * @param _elems Bytes32 array to be hashed.
      *
-     * @return A hash of element.
+     * @return A hash of elements tree.
      */
     function _hashElems(bytes32[] memory _elems) internal view returns (bytes32) {
-        require(_elems.length >= 2, "ZKTrieHasher: too few values for _hashElems");
-        bytes32 baseH = POSEIDON2.poseidon([_elems[0], _elems[1]]);
-        if (_elems.length == 2) return baseH;
-        else if (_elems.length == 3) {
-            return POSEIDON2.poseidon([baseH, _elems[2]]);
-        }
-        bytes32[] memory newValues = new bytes32[](_elems.length / 2 + 1);
-        newValues[0] = baseH;
-        for (uint256 i = 0; i < newValues.length; ) {
-            if ((i + 1) * 2 > _elems.length) {
-                newValues[i] = _elems[i * 2 + 1];
-            } else {
-                newValues[i] = POSEIDON2.poseidon([_elems[i * 2], _elems[i * 2 + 1]]);
-            }
-            unchecked {
-                ++i;
+        require(_elems.length >= 4, "ZKTrieHasher: too few values for _hashElems");
+        IPoseidon2 iposeidon = POSEIDON2;
+
+        uint256 idx;
+        uint256 adjacent_idx;
+
+        uint256 adjacent_offset = 1;
+        uint256 jump = 2;
+        uint256 length = _elems.length;
+        for (; adjacent_offset < length;) {
+            for (idx = 0; idx < length;) {
+                unchecked {
+                    adjacent_idx = idx + adjacent_offset;
+                }
+                if (adjacent_idx < length) {
+                    _elems[idx] = iposeidon.poseidon( [_elems[idx], _elems[adjacent_idx]] );
+                }
+                unchecked {
+                    idx += jump;
+                }
             }
+            adjacent_offset = jump;
+            jump <<= 1;
         }
-        return _hashElems(newValues);
+
+        return _elems[0];
+    }
+
+    /**
+     * @notice Computes a root hash of 2 elements.
+     *
+     * @param left_leaf  Bytes32 left leaf to be hashed.
+     * @param right_leaf Bytes32 right leaf to be hashed.
+     *
+     * @return A hash of 2 elements.
+     */
+    function _hashFixedElems2(bytes32 left_leaf, bytes32 right_leaf) internal view returns (bytes32) {
```

Comment:
How about changing the function name to `_hashFixed2Elems`, `_hashFixed3Elems`? I think it's more understandable.

Reply:
renamed.